plugins/permission/src/components/PermissionedRoute.tsx

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { ComponentProps, ReactElement } from 'react';
+import { Route } from 'react-router';
+import { useApp } from '@backstage/core-plugin-api';
+import { usePermission } from '../hooks';
+import { Permission } from '@backstage/plugin-permission-common';
+
+/**
+ * Returns a React Router Route which only renders the element when authorized. If unathorized, the Route will render a
+ * {@link NotFoundErrorPage}.
+ * @public
+ */
+export const PermissionedRoute = ({
+  permission,
+  resourceRef,
+  ...props
+}: ComponentProps<typeof Route> & {
+  permission: Permission;
+  resourceRef?: string;
+}) => {
+  const permissionResult = usePermission(permission, resourceRef);
+  const app = useApp();
+  const { NotFoundErrorPage } = app.getComponents();
+
+  let shownElement: ReactElement | null | undefined = <NotFoundErrorPage />;
```

Comment:
Plugins may want to customize the unauthorized / loading views here. Thoughts on adding those as optional props with the current elements as a fallback? Same feedback on `WithPermission`

Reply:
While I feel like plugins should probably defer to showing an error page provided by their backstage app for the sake of consistency, I'm on board with this idea - I'm sure there are cases where this may be useful.